implementations/rust/ockam/ockam_transport_udp/src/workers/listener.rs

```diff
@@ -61,32 +60,41 @@ impl Processor for UdpListenProcessor {
     }
 
     async fn process(&mut self, ctx: &mut Self::Context) -> Result<bool> {
-        debug!("Waiting for incoming UDP datagram...");
+        debug!(
+            "Waiting for incoming UDP datagram on {}...",
+            self.local_addr
+        );
         let (mut msg, addr) = match self.stream.next().await {
             Some(res) => match res {
                 Ok((msg, addr)) => (msg, addr),
-                Err(_e) => {
-                    info!("Failed to read message from UDP socket.");
+                Err(e) => {
+                    warn!("Failed to read message from UDP socket: {:?}", e);
+
+                    // Inform router of error
+                    let _ = self
+                        .router_handle
+                        .sender_listener_error(self.local_addr)
```

Comment:
I'm not sure re-building the UDP socket when it breaks it's the way forward:
> Can put the sending part of a socket into an error state, for example, by trying to send to an IPv6 address.

Can't we avoid breaking the socket by allowing only ipv4? Are there other cases?

Reply:
I've seen the sink part of a socket get in to an error state when (1) sending to an IPv6 address and (2) when sending to "0.0.0.0:0". The transport does not support IPv6, so (1) can no longer happen, but something like (2) or something I've never seen could happen.

My gut feeling is that errors will be rare as UDP is happy to send to peers that have no route or are disconnected. However, I don't know what errors could happen on all platforms and now, as the transport uses a single UDP socket for 'client' messages to all peers, if an error occurs the transport will grind to a halt for 'client' messages unless the transport tries to clear/recover from the error.

I'm happy to implement whatever Ockam prefer. 👍

Reply:
> as the transport uses a single UDP socket for 'client' messages to all peers, if an error occurs the transport will grind to a halt for 'client' messages unless the transport tries to clear/recover from the error.

I 100% agree, we cannot afford to have a broken socket when the whole node depends on it.
Still, what worries me most, is the scenario where one worker keeps re-trying sending a message and keeps breaking the socket. In this case, the service seems to be working fine and will probably go undetected, but the packet drop is higher than it should be.

> However, I don't know what errors could happen on all platforms

How about fixing errors one step at the time? I believe a broken socket doesn't go unnoticed, and will help identify and fix every failure case in every platform.

If you could please reject and log messages sent toward `0.0.0.0`, we could omit the recover mechanism.

Reply:
Okay, cool 👍 .

I've added the below check in `UdpSendWorker` just before sending, and have removed the recover mechanism.

```
        // Error on conditions that _might_ put the sink
        // into an error state
        if addr.port() == 0 {
            warn!(peer_addr = %peer_addr, "Will not send to address");
            return Err(TransportError::InvalidAddress.into());
        }
```

From testing, it seems sending to a port 0 causes the sink to get into an error state. For example, sending to `0.0.0.0:8080` is okay but `0.0.0.0:0` and `192.168.1.10:0` are not.